cupy/cuda/stream.pyx

```diff
@@ -178,32 +203,42 @@ class BaseStream(object):
         # This operator is implemented to compare the singleton instance
         # of null stream (Stream.null) can safely be compared with null
         # stream instance created by a user.
-        return self.ptr == other.ptr
+        if self.dev == -1 or other.dev == -1:
+            return self.ptr == other.ptr
+        else:
+            return (self.dev == other.dev and self.ptr == other.ptr)
 
     def __enter__(self):
         tls = _ThreadLocal.get()
-        if tls.prev_stream_ref_stack is None:
-            tls.prev_stream_ref_stack = []
-        prev_stream_ref = tls.get_current_stream_ref()
-        tls.prev_stream_ref_stack.append(prev_stream_ref)
+        cdef int dev = self.dev
+        dev = check_stream_device_match(dev)
+        # to prevent from popping the wrong stream at exit
+        self._curr_dev = dev
```

Comment:
As a `BaseStream` instance may `__enter__` recursively (regardless if it makes sense), we also have to manage `_curr_dev` in the stack. Pushing `(dev, prev_stream_ref)` instead of `prev_stream_ref` only and poping it would be enough.

```Python
with cuda.Device(0):
    s = cuda.Stream()
    with s:
        with cuda.Device(1):
            with s:  # <--- s._curr_dev is replaced with 1
                ...
        ...
        # The first context `__exit__`s with popping the wrong stream
```

Reply:
(maybe nitpick)

Reply:
Your snippet will raise an error with this PR, because I added a `check_stream_device_match` check. However, I haven't thought about this possibility (which is clearly an abuse that should not happen...):
```python 
with cuda.Device(0):
    s = cuda.Stream()
    with s:
        with s:  # <--- s._curr_dev is washed out
            ...
        ...
```
@takagi should we worry about this?

Reply:
Please leave it as it is. I was trapped by a minor thing. (I initially thought of making the interpreter's context stack manage the current stream history, but given `Stream.null` and `Stream.ptds`, I gave it up as they are shared between threads and we eventually need thread-local data management unless we provide the instances (`null` and `ptds`) for each thread)

Reply:
I realized accessing `self._curr_dev` is unsafe either for precisely the same reason (the singletons are not thread-safe), but your suggestion 

> Pushing `(dev, prev_stream_ref)` instead of `prev_stream_ref` only and poping it would be enough.

cannot be taken as is, because we need to know `dev` to pop from the right stack. We need to keep track of two stacks to address this. Working on it...

Reply:
Fixed in 9cdd026.